### PR TITLE
PLANNER-2672 Return CompletableFuture for a problem change

### PR DESCRIPTION
**Pull request by rsynek**
TODO:
- [x] check and update documentation

### JIRA
https://issues.redhat.com/browse/PLANNER-2672

### Referenced pull requests
https://github.com/kiegroup/optaplanner-website/pull/485

### Checklist
- [ ] Documentation updated if applicable.
- [ ] Upgrade recipe provided if applicable.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>specific pull request build</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] tests</b>
* for a <b>full downstream build</b> 
  * for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  * for <b>github actions</b> job: add the label `run_fdb`
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
* for a <b>Quarkus LTS check</b> please add comment: <b>Jenkins run LTS</b>
* for a <b>specific Quarkus LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] LTS</b>
* for a <b>Native check</b> please add comment: <b>Jenkins run native</b>
* for a <b>specific Native LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] native</b>
</details>

### CODE REVIEW

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/api/solver/SolverJob.java [hunk 1]
```diff
@@ -55,10 +56,11 @@
      * Schedules a {@link ProblemChange} to be processed by the underlying {@link Solver} and returns immediately.
      *
      * @param problemChange never null
+     * @return CompletableFuture<Void> that completes after the best solution containing this change has been consumed.
```
Comment (triceo) [inline core/optaplanner-core-impl/src/main/java/org/optaplanner/core/api/solver/SolverJob.java:59]:
```suggestion
     * @return completes after the best solution containing this change has been consumed.
```

Don't see the need for duplicating the type information which is already there.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/api/solver/SolverManager.java [hunk 2]
```diff
@@ -312,9 +313,10 @@
      * @param problemId never null, a value given to {@link #solve(Object, Function, Consumer)}
      *        or {@link #solveAndListen(Object, Function, Consumer)}
      * @param problemChange never null
+     * @return CompletableFuture<Void> that completes after the best solution containing this change has been consumed.
```
Comment (triceo) [inline core/optaplanner-core-impl/src/main/java/org/optaplanner/core/api/solver/SolverManager.java:316]:
```suggestion
     * @return completes after the best solution containing this change has been consumed.
```

Dtto.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        }
+        SortedMap<BigInteger, List<CompletableFuture<Void>>> containedProblemChangesPerVersion =
+                problemChangesPerVersion.headMap(versionedBestSolution.getVersion().add(BigInteger.ONE));
+
+        List<CompletableFuture<Void>> containedProblemChanges = containedProblemChangesPerVersion.values()
+                .stream()
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+
+        containedProblemChangesPerVersion.keySet().forEach(problemChangesPerVersion::remove);
```
Comment (triceo) [inline core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java:63]: Why not just do it in one `forEach(...)`? I see the intention for separation of concerns, but this code is simple enough that I think the double iteration is not worth it.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+import org.optaplanner.core.api.solver.Solver;
+import org.optaplanner.core.api.solver.change.ProblemChange;
+
+final class BestSolutionHolder<Solution_> {
+
+    private final Lock problemChangesLock = new ReentrantLock();
+    private final AtomicReference<VersionedBestSolution> versionedBestSolutionRef = new AtomicReference<>();
+    private final SortedMap<BigInteger, List<CompletableFuture<Void>>> problemChangesPerVersion =
+            new TreeMap<>();
```
Comment (triceo) [inline core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java:40]: Can you guarantee that `take()` is never called concurrently?
Because the logic in there pretty much guarantees CMEs if it is.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+    /**
+     * Adds a new problem change to a solver and registers the problem change to be later retrieved together with
+     * a relevant best solution by the {@link #take()} method.
+     * 
+     * @return CompletableFuture that will be completed after the best solution containing this change is passed to
+     *         a user-defined Consumer.
+     */
+    CompletableFuture<Void> addProblemChange(Solver<Solution_> solver, ProblemChange<Solution_> problemChange) {
+        CompletableFuture<Void> futureProblemChange;
```
Comment (triceo) [inline core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java:102]: Why is this here? You can easily return from the try block directly.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                    .stream()
+                    .flatMap(Collection::stream)
+                    .forEach(pendingProblemChange -> pendingProblemChange.cancel(false));
+            problemChangesPerVersion.clear();
+        } finally {
+            problemChangesLock.unlock();
+        }
+    }
+
+    private final class VersionedBestSolution {
```
Comment (triceo) [inline core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java:134]: I'd make it static. You'll have to reintroduce `<Solution_>`, but that's not an issue, I think.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java [hunk 7]
```diff
@@ -97,15 +108,20 @@ private void tryConsumeWaitingIntermediateBestSolution() {
...
+                try {
+                    bestSolutionConsumer.accept(bestSolutionContainingProblemChanges.getBestSolution());
+                } catch (Throwable throwable) {
+                    exceptionHandler.accept(problemId, throwable);
+                } finally {
+                    /*
+                     * Complete the CompletableFutures associated with ProblemChanges regardless of a possible
+                     * exception in the consumer to avoid a memory leak.
+                     */
+                    bestSolutionContainingProblemChanges.completeProblemChanges();
```
Comment (triceo) [inline core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java:122]: Shouldn't those futures be completed exceptionally in case of an exception?

**Review by triceo — COMMENTED** [review]
I'm leaving some questions.
I will approve once we have good answers.

**Review by yurloc — COMMENTED** [review]
Suggestions.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java [hunk 8]
```diff
@@ -20,43 +20,46 @@
...
 
     public ConsumerSupport(ProblemId_ problemId, Consumer<? super Solution_> bestSolutionConsumer,
             Consumer<? super Solution_> finalBestSolutionConsumer,
-            BiConsumer<? super ProblemId_, ? super Throwable> exceptionHandler) {
+            BiConsumer<? super ProblemId_, ? super Throwable> exceptionHandler,
+            BestSolutionHolder<Solution_> bestSolutionHolder) {
         this.problemId = problemId;
         this.bestSolutionConsumer = bestSolutionConsumer;
-        this.finalBestSolutionConsumer = finalBestSolutionConsumer;
+        this.finalBestSolutionConsumer = finalBestSolutionConsumer == null ? finalBestSolution -> {
```
[the remark this answers is not in the document]
Comment (rsynek) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java:43]: Thanks for the tip; however, to me, it doesn't look better:
`Objects.requireNonNullElseGet(finalBestSolutionConsumer, () -> finalBestSolution -> {});`
Especially passing a supplier of a consumer as the second argument is not that straightforward.

File: core/optaplanner-core-impl/src/test/java/org/optaplanner/core/api/solver/SolverManagerTest.java [hunk 9]
```diff
@@ -609,26 +603,21 @@ void addProblemChangeToWaitingSolver() throws InterruptedException {
         // The second solver is scheduled and waits for the fist solver to finish.
         final long secondProblemId = 2L;
         final int entityAndValueCount = 4;
-        CountDownLatch solutionWithProblemChangeReceived = new CountDownLatch(1);
+        // CountDownLatch solutionWithProblemChangeReceived = new CountDownLatch(1);
```
[the remark this answers is not in the document]
Comment (rsynek) [reply core/optaplanner-core-impl/src/test/java/org/optaplanner/core/api/solver/SolverManagerTest.java:606]: Good catch, thanks.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java (hunk 4 above)
Comment (rsynek) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java:40]: Yes, this method is always called only from the single consumer thread. But I will make it clear in the Javadoc that the method is not thread-safe.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java (hunk 6 above)
Comment (rsynek) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java:134]: Good point.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java (hunk 7 above)
Comment (triceo) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java:122]: Hmmm. You are right that the user can request the exception handler. But is the user *required* to? (And if not, what happens if he does not provide one? Is the exception just swallowed?)

Comment (rsynek) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java:122]: No, if the exception handler is not provided, a default one is used, which just logs the exception. Not ideal, though.

Comment (triceo) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java:122]: Then maybe we should complete exceptionally in that case? Or is that too complicated already?

Comment (rsynek) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java:122]: In the end, I decided to complete the future exceptionally in the case described above. If the user is waiting for the best solution, it's safer to make sure the exception in the consumer will not be missed.

**Review by triceo — APPROVED** [review]
LGTM.
@rsynek please update N&N and upgrade recipe, as this is theoretically backwards incompatible.

**Comment by rsynek** [comment]
jenkins retest this
